prov/verbs/src/verbs_domain_xrc.c

```diff
@@ -0,0 +1,576 @@
+	if (ret) {
+		ep->ini_conn->state = last_state;
+		fi_ibv_put_shared_ini_conn(ep);
+
+		/* We need to let the application know that the connect request
+		 * has failed. */
+		fi_ibv_create_shutdown_event(ep);
+	}
+
+	/* Continue to schedule shared connections if the physical connection
+	 * has completed and there are connection requests pending. We could
+	 * implement a throttle here if it is determined that it is better to
+	 * limit the number of outstanding connections. */
+	goto sched_next;
+
+	return;
+}
+
+/* Caller must hold domain:xrc:ini_mgmt_lock */
+int fi_ibv_process_ini_conn(struct fi_ibv_ep *ep,int reciprocal,
+			    void *param, size_t paramlen)
+{
+	struct fi_ibv_xrc_cm_data *cm_data = param;
+	struct rdma_conn_param conn_param = { 0 };
+	int ret;
+
+	assert(ep->ibv_qp);
+
+	if (!reciprocal)
+		fi_ibv_eq_set_xrc_conn_tag(ep);
+
+	fi_ibv_set_xrc_cm_data(cm_data, reciprocal, ep->conn_setup->tag,
+			       ep->eq->pep_port, ep->ini_conn->tgt_qpn);
+	conn_param.private_data = cm_data;
+	conn_param.private_data_len = paramlen;
+	conn_param.responder_resources = RDMA_MAX_RESP_RES;
+	conn_param.initiator_depth = RDMA_MAX_INIT_DEPTH;
+	conn_param.flow_control = 1;
+	conn_param.retry_count = 15;
+	conn_param.rnr_retry_count = 7;
+	conn_param.srq = 1;
+
+	/* Shared connections use reserved temporary QP numbers to
+	 * avoid the appearance of stale/duplicate CM messages */
+	if (!ep->id->qp)
+		conn_param.qp_num = ep->conn_setup->rsvd_ini_qpn->qp_num;
+
+	assert(ep->conn_state == FI_IBV_XRC_UNCONNECTED ||
+	       ep->conn_state == FI_IBV_XRC_ORIG_CONNECTED);
+	ep->conn_state++;
+
+	ret = rdma_connect(ep->id, &conn_param) ? -errno : 0;
+	if (ret) {
+		ret = -errno;
+		VERBS_INFO_ERRNO(FI_LOG_FABRIC, "rdma_connect", errno);
+		ep->conn_state--;
+	}
+	return ret;
+}
+
+int fi_ibv_ep_create_tgt_qp(struct fi_ibv_ep *ep, uint32_t tgt_qpn)
+{
+	struct ibv_qp_open_attr open_attr;
+	struct ibv_qp_init_attr_ex attr_ex;
+	struct fi_ibv_domain *domain = fi_ibv_msg_ep_to_domain(ep);
+	struct ibv_qp *rsvd_qpn;
+
+	assert(ep->tgt_id && !ep->tgt_id->qp);
+
+	/* If a target QP number was specified then open that existing
+	 * QP for sharing. */
+	if (tgt_qpn) {
+		rsvd_qpn = fi_ibv_reserve_qpn(ep);
+		if (!rsvd_qpn) {
+			VERBS_WARN(FI_LOG_FABRIC,
+				   "Create of XRC reserved QPN failed %d\n",
+				   errno);
+			return -errno;
+		}
+
+		memset(&open_attr, 0, sizeof(open_attr));
+		open_attr.qp_num = tgt_qpn;
+		open_attr.comp_mask = IBV_QP_OPEN_ATTR_NUM |
+			IBV_QP_OPEN_ATTR_XRCD | IBV_QP_OPEN_ATTR_TYPE |
+			IBV_QP_OPEN_ATTR_CONTEXT;
+		open_attr.xrcd = domain->xrc.xrcd;
+		open_attr.qp_type = IBV_QPT_XRC_RECV;
+		open_attr.qp_context = ep;
+
+		ep->tgt_ibv_qp = ibv_open_qp(domain->verbs, &open_attr);
+		if (!ep->tgt_ibv_qp) {
+			VERBS_INFO_ERRNO(FI_LOG_EP_CTRL,
+				   "XRC TGT QP, ibv_open_qp()", errno);
+			fi_ibv_release_qpn(rsvd_qpn);
+			return -errno;
+		}
+		ep->conn_setup->rsvd_tgt_qpn = rsvd_qpn;
+		return FI_SUCCESS;
```

Comment:
drop return and add else clause given the structure of this function -- or split into 2 functions and have caller invoke based on tgt_qpn

Reply:
Done.